File: leetcode/editor/cn/345_反转字符串中的元音字母_test.go
package leetcode

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func Test_givenNormal_when345_thenSuccess(t *testing.T) {
	assert.Equal(t, 1, 1, "shouldEqual")
}

//leetcode submit region begin(Prohibit modification and deletion)
func reverseVowels(s string) string {
	void := struct{}{}
	vowels := map[byte]struct{}{
		'a': void, 'e': void, 'i': void, 'o': void, 'u': void,
		'A': void, 'E': void, 'I': void, 'O': void, 'U': void,
	}
	// TODO string to bytes
	bytes := []byte(s)
	for left, right := 0, len(bytes)-1; left < right; {
		if _, ok := vowels[bytes[left]]; !ok {
			left += 1
			continue
		}
		if _, ok := vowels[bytes[right]]; !ok {
			right -= 1
			continue
		}
		bytes[left], bytes[right] = bytes[right], bytes[left]
		left, right = left+1, right-1
	}
	return string(bytes)
}

//leetcode submit region end(Prohibit modification and deletion)

File: leetcode/editor/cn/345_反转字符串中的元音字母_test.py
import unittest


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        self.solution.reverseVowels("Hello")
        self.assertTrue(True, "ShouldBeTrue")


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def reverseVowels(self, s: str) -> str:
        # TODO str => list[str]
        split: list[str] = list(s)
        vowels = {
            'a', 'e', 'i', 'o', 'u',
            'A', 'E', 'I', 'O', 'U'
        }
        left, right = 0, len(s) - 1
        while left < right:
            if split[left] not in vowels:
                left += 1
                continue
            if split[right] not in vowels:
                right -= 1
                continue
            split[left], split[right] = split[right], split[left],
            left, right = left + 1, right - 1
        # TODO list to str
        return ''.join(split)


# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()
